src/main/scala/strawman/collection/immutable/ListSet.scala

```diff
@@ -0,0 +1,126 @@
+package strawman
+package collection.immutable
+
+import strawman.collection.mutable.{Builder, ImmutableSetBuilder}
+import strawman.collection.IterableFactory
+
+import scala.annotation.tailrec
+import scala.{Any, Boolean, Int, NoSuchElementException, SerialVersionUID, Serializable}
+
+/**
+  * $factoryInfo
+  *
+  * Note that each element insertion takes O(n) time, which means that creating a list set with
+  * n elements will take O(n^2^) time. This makes the builder suitable only for a small number of
+  * elements.
+  *
+  * @since 1
+  * @define Coll ListSet
+  * @define coll list set
+  */
+object ListSet extends IterableFactory[ListSet] {
+
+  def newBuilder[A <: Any]: Builder[A, ListSet[A]] = new ImmutableSetBuilder[A, ListSet](empty[A])
+
+  def fromIterable[E](it: strawman.collection.Iterable[E]): ListSet[E] = newBuilder.++=(it).result
+
+  @SerialVersionUID(5010379588739277132L)
+  private object EmptyListSet extends ListSet[Any]
+  private[collection] def emptyInstance: ListSet[Any] = EmptyListSet
+
+}
+
+/**
+  * This class implements immutable sets using a list-based data structure. List set iterators and
+  * traversal methods visit elements in the order whey were first inserted.
+  *
+  * Elements are stored internally in reversed insertion order, which means the newest element is at
+  * the head of the list. As such, methods such as `head` and `tail` are O(n), while `last` and
+  * `init` are O(1). Other operations, such as inserting or removing entries, are also O(n), which
+  * makes this collection suitable only for a small number of elements.
+  *
+  * Instances of `ListSet` represent empty sets; they can be either created by calling the
+  * constructor directly, or by applying the function `ListSet.empty`.
+  *
+  * @tparam A the type of the elements contained in this list set
+  *
+  * @author Matthias Zenger
+  * @version 1.0, 09/07/2003
+  * @since 1
+  * @define Coll ListSet
+  * @define coll list set
+  * @define mayNotTerminateInf
+  * @define willNotTerminateInf
+  */
+@SerialVersionUID(-8417059026623606218L)
+sealed class ListSet[A]
+  extends Set[A]
+    with SetLike[A, ListSet]
+    with Serializable {
+
+  override def size: Int = 0
+  override def isEmpty: Boolean = true
+
+  def contains(elem: A): Boolean = false
+
+  def + (elem: A): ListSet[A] = new Node(elem)
+  def - (elem: A): ListSet[A] = this
+
+  def iterator(): strawman.collection.Iterator[A] = {
+    def reverseList = {
```

Comment:
It usually ends up as an extra method in the bytecode.  Sometimes it's faster because each method is below some important limit for inlining or optimization.  Sometimes it's slower because the extra method call either isn't elided or is but uncleanly leaving extra work to do.  Most of the time it doesn't matter.

But the thing that I don't understand is why it is even here.  The exact same algorithm with the exact same variables would be used if there wasn't a `def`.  So it seems a pointless extra layer.